rust/tools/authorization-logic/src/souffle/datalog_ir.rs

```diff
@@ -18,14 +18,20 @@
 //! this authorization logic into datalog simpler.
 use crate::ast::*;
 
+#[derive(Clone)]
+pub enum DLIRRvalue {
+    PredicateRvalue { predicate: AstPredicate },
+    BinopRValue { lnum: String, binop: AstBinop, rnum: String }
```

Comment:
It looks like this alternative is the same as the one in `AstRValue`. Should the `BinopRValue` definition be extracted and shared?

Reply:
The `BinOpRvalue` branch is the same yes, but the overall definition of `RValue` is different because here the other branch is a predicate but in the other case it's a flatfact. 

I guess it could be split into a separate class, but I think other parts of the code would become more verbose. I also think there is some merit to going the other way and making the other class which is shared between them (`AstPredicate`) distinct because then the layers of the IR are totally separate and then you get type errors if you try to run a compiler pass that was meant for one layer of the IR on another.

Reply:
I've looked around, and I don't think it would be that much more verbose. I think it would end up saving space if you end up adding more structure under `lnum` and `rnum`. That said, none of this code is going to live very long, so I'll defer to your judgement on this one.

Reply:
fixed